src/shogun/mathematics/linalg/internal/implementation/Mean.h

```diff
@@ -0,0 +1,984 @@
+{
+	/** Scalar type */
+	typedef typename Matrix::Scalar T;
+
+	/** Return type */
+	typedef CGPUVector<T> ReturnType;
+
+	/** Generates the computation kernel */
+	template <class T>
+	static viennacl::ocl::kernel& generate_kernel(bool no_diag)
+	{
+		std::string kernel_name = "colwise_mean_" + ocl::get_type_string<T>();
+		if (no_diag) kernel_name.append("_no_diag");
+
+		if (ocl::kernel_exists(kernel_name))
+			return ocl::get_kernel(kernel_name);
+
+		std::string source = ocl::generate_kernel_preamble<T>(kernel_name);
+		if (no_diag) source.append("#define NO_DIAG\n");
+
+		source.append(
+			R"(
+				__kernel void KERNEL_NAME(
+					__global DATATYPE* mat, int nrows, int ncols, int offset,
+					__global DATATYPE* result, int result_offset)
+				{
+					int j = get_global_id(0);
+
+					if (j>=ncols)
+						return;
+
+					DATATYPE mean = 0;
+					for (int i=0; i<nrows; i++)
+					{
+					#ifdef NO_DIAG
+						if (i!=j)
+					#endif
+						mean += mat[offset+i+j*nrows];
+					}
+
+					result[j+result_offset] = mean;
+				}
+			)"
+		);
+
+		viennacl::ocl::kernel& kernel = ocl::compile_kernel(kernel_name, source);
+
+		kernel.local_work_size(0, OCL_WORK_GROUP_SIZE_1D);
+
+		return kernel;
+	}
+
+	/**
+	 * Method that computes the column wise mean of CGPUMatrix using ViennaCL
+	 *
+	 * @param m the matrix whose colwise mean has to be computed
+	 * @param no_diag if true, diagonal entries are excluded from the mean
+	 * @return the colwise mean computed as \f$s_j=\mean_{i}m_{i,j}\f$
+	 */
+	static CGPUVector<T> compute(CGPUMatrix<T> m, bool no_diag)
+	{
+                REQUIRE(m.num_rows > 0, "Matrix can not be empty!\n");
+                CGPUVector<T> result(m.num_cols);
+		compute(m, result, no_diag);
+		return result;
+	}
+
+	/**
+	 * Method that computes the column wise mean of CGPUMatrix blocks
+	 * using ViennaCL
+	 *
+	 * @param b the matrix-block whose colwise mean has to be computed
+	 * @param no_diag if true, diagonal entries are excluded from the mean
+	 * @return the colwise mean computed as \f$s_j=\mean_{i}b_{i,j}\f$
+	 */
+	static CGPUVector<T> compute(Block<CGPUMatrix<T> > b, bool no_diag)
+	{
+		SG_SERROR("The operation colwise_mean() on a matrix block is currently not supported\n");
+		return CGPUVector<T>();
+	}
+
+	/**
+	 * Method that computes the column wise mean of CGPUMatrix using ViennaCL
+	 *
+	 * @param m the matrix whose colwise mean has to be computed
+	 * @param no_diag if true, diagonal entries are excluded from the mean
+	 * @param result Pre-allocated vector for the result of the computation
+	 */
+	static void compute(CGPUMatrix<T> mat, CGPUVector<T> result, bool no_diag)
+	{
+                CGPUVector<T> sum_result(mat.num_cols);
+                sum_result = colwise_sum<Backend::VIENNACL, CGPUMatrix<T> >::compute(mat, no_diag);
+                if (no_diag) 
+                {
+                        REQUIRE(mat.num_rows != mat.num_cols, "Matrix is not square!\n");
```

Comment:
Please in error messages always give full information.
Matrix has size %dx%d, which is not square
See the dev guidelines on this here:
https://github.com/shogun-toolbox/shogun/wiki/Assertions